File: src/org/vaadin/multiprobe/data/Probe.java
package org.vaadin.multiprobe.data;

import java.io.IOException;
import java.io.Writer;

public class Probe {
    private final ProbeData analData;

    private ProbeEntry data;
    private ProbeEntry tail;
    private int count = 0;
    
    private int id;
    public Probe(ProbeData analData, int id) {
        this.analData = analData;
        this.id = id;
    }
    
    public void add(ProbeEntry entry, Writer journal) {
        if (data == null)
            data = tail = entry;
        else {
            tail.setNext(entry);
            tail = entry;
        }
        count++;

        if (journal != null)
            try {
                entry.write(id, journal);
            } catch (IOException e) {
                System.err.println("Could not write entry to journal");
            }
    }
    
    public int getId() {
        return id;
    }
    
    public ProbeEntry getData() {
        return data;
    }
    
    public int size() {
        return count;
    }
}

File: src/org/vaadin/multiprobe/data/Probe.scala
package org.vaadin.multiprobe.data

import java.util.Date
import scala.collection.mutable.LinkedList
import java.io.Writer
import java.io.IOException
import java.util.Observable

/**
 * Describes one instrument attached to the collector device.
 */
class Probe(val id: Int) extends Observable {
  var data: LinkedList[ProbeEntry] = LinkedList()
  var afterLast: LinkedList[ProbeEntry] = data
  var last : LinkedList[ProbeEntry] = null

  def add(entry: ProbeEntry, journal: Writer) = {
    afterLast.elem = entry
    afterLast.next = LinkedList()
    last = afterLast
    afterLast = afterLast.next
    
    if (journal != null)
      try {
        entry.write(id, journal)
      } catch {
        case e: IOException =>
          System.err.println("Could not write entry to journal");
      }
  }
  
  override def toString: String = {
    if (data.isEmpty)
      "{}"
    else
      "{" + data.map(entry => "(" + entry.time + "," + entry.value + ")").reduceLeft((a,b) => a + ", " + b) + "}"
  }
}
